### ISSUE TITLE
Question: C# API Inferencing of Tabular Data (1D Tensors) with Different Dimensions

### ISSUE THREAD

**Issue by whitmark** [microsoft/onnxruntime#5846]
**Describe the bug**
Looking to update C# code to accommodate latest ML.net and onnxruntime versions.  Getting hung-up on a similar issue as my original post [#1417](https://github.com/microsoft/onnxruntime/issues/1417) that was working under a previous onnxruntime v0.4.0 but is no longer under v1.5.2.

**Urgency**
On critical path to update production code and operations.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Win 10 Pro 19041
- ONNX Runtime installed from (source or binary):  binary
- ONNX Runtime version:  1.5.2
- ONNX Runtime
- Visual Studio version (if applicable):  16.8.1
- Microsoft.ML.OnnxConverter: 0.17.2

**To Reproduce**
When building the model, I have 100 input columns of which 1 is a binary Label (0,1) and 96 are used as Features with the remaining 3 are reference columns I can use to identify the source of the record.  When I attempt to inference a test record with one dummy label column and 96 feature values, I get an unhandled exception when I attempt to assign the dense tensor: 

> System.ArgumentOutOfRangeException: 'Dimensions must be non-negative Parameter name: dimensions'

See attached image below.

- Describe steps/code to reproduce the behavior.
```csharp
 Console.WriteLine("=============== Inference Onnx Model Begin ===============");

 // Optional : Create session options and set the graph optimization level for the session
 SessionOptions sessOptions = new SessionOptions();
 sessOptions.GraphOptimizationLevel = GraphOptimizationLevel.ORT_ENABLE_BASIC;

 using (var session = new InferenceSession(onnxModelPath, sessOptions))
 {
        var inputMeta = session.InputMetadata;
        var container = new List<NamedOnnxValue>();

         float[] inputData1 = new float[97] { 1f,   // Dummy Label Column
                                            35f, 24f, 13.4883f, 25f, 99f, 62.7967f, 48f, 47f, 28.4488f, 60f,
                                            75f, 49.3083f, 19f, 11f, 12.4294f, 12f, 28f, 20.8595f, 31f, 17f,
                                            8.4301f, 13f, 6f, 4.718f, 33.25f, 133f, 99f, 62.7967f, 41.0069f, 35.8817f,
                                            135.1481f, 133.6822f, 0.71f, 1.37f, 1.71f, 0.54f, 0.4f, 0.65f, 1.63f, 2.58f,
                                            4.13f, 1.96f, 3.13f, 0.46f, 0.23f, 0.36f, 1.55f, 0.61f, 4.66f, 2.11f,
                                            3.66f, 0.44f, 0.92f, 0.3f, 0.68f, 0.4f, 0.5833f, 0f, 0.8f, 0.4833f,
                                            1f, 0.7833f, 0f, 0.2286f, 0.6762f, 0.781f, 0.9429f, 1f, 0f, 0.1998f,
                                            0.6212f, 0.8053f, 0.9301f, 1f, 0.4177f, 0.4271f, 0.4469f, 0.4555f, 0.4613f, 0.4646f,
                                            0.2025f, 0.2039f, 0.2067f, 0.2079f, 0.2088f, 0.2092f, 0.5125f, 0.5112f, 0.5276f, 0.5574f,
                                            0.5458f, 0.698f, 1.1195f, 1.1928f, 0.9797f, 1.01f };

        int i = 0;
        int startSlot = 3;
        foreach (var name in inputMeta.Keys)
        {
           if (i >= startSlot) // Skip reference Fields except rHProf
            { 
                Console.WriteLine(name);
                var tensor = new DenseTensor<float>(new float[] { inputData1[i - startSlot] }, inputMeta[name].Dimensions);
                 container.Add(NamedOnnxValue.CreateFromTensor<float>(name, tensor));
            }
            i++;
        }

        // Run the inference
        // results is an IDisposableReadOnlyCollection<DisposableNamedOnnxValue> container
        using (var results = session.Run(container))  
        {
              // dump the results
               foreach (var r in results)
               {
                     Console.WriteLine(string.Format("Output for {0} = ", r.Name));
                     Console.WriteLine(string.Format("{0}", r.AsTensor<float>().GetArrayString()));
                }
         }
  }

    Console.WriteLine("=============== Inference Onnx Model End ===============");
}
catch (Exception ex)
{
    Console.WriteLine("Error: {0}\nCallStack: {1}\n Inner Exception: {2}", ex.Message, ex.StackTrace, ex.InnerException != null ? ex.InnerException.Message : "No Inner Exception");
    throw ex;
}

```
- Attach the ONNX model to the issue (where applicable) to expedite investigation.
[MES_4Range_UpFPP_ZZ16_HP12_20200101_20201031_ONNX_Model.zip](https://github.com/microsoft/onnxruntime/files/5559863/MES_4Range_UpFPP_ZZ16_HP12_20200101_20201031_ONNX_Model.zip)

**Expected behavior**
Looking for a straightforward way to inference an onnx model when the dimensions of the training/testiing data, features, and inferencing data do not precisely line up.  Model building is an iterative process whereby I am constantly performing different experiments by adding/removing features and would prefer not to have to build or modify a schema each time if I can properly manage the sequence of feature columns within the block between training/testing and inferencing.  Plus I will always want to maintain a few reference data items in the first few columns of each record to know the source of the data.    

**Screenshots**

Unhandled Exception:
![Snag_5045147](https://user-images.githubusercontent.com/8396992/99527705-3d342880-296b-11eb-9415-4970bb992fa4.png)

Inference Session inputMetaData:
![Snag_5049322](https://user-images.githubusercontent.com/8396992/99527734-491fea80-296b-11eb-9112-ef8f57fa297a.png)

**Additional context**

**Comment by yuslepukhin** [comment]
Your input shape must match the one that model expects. You model has dynamic shape inputs. -1 indicates the dimension that can vary input with input, however, you certainly can not feed -1 with your input since the code is not able to calculate the buffer length you are passing. For example, for the input shape [-1,1] that you are having you can change the first dim with every input.
For example, [1,1], [2,1] etc.

**Comment by yuslepukhin** [comment]
BTW, version 0.4 is certainly not a previous version.

**Comment by whitmark** [comment]
Thanks @yuslepukhin.  I see now the Dimensions on each InputMetaData item is [-1,1] but unclear why the model trained with one-dimensional tabular data has dynamic shape inputs.   The model was trained using ML.Net LightGbm and converted to onnx. 

```
using (var file = File.Create(ModelOnnxPath))
       mlContext.Model.ConvertToOnnx(mlModel, trainingDataView, file);
```

As you can see from my post [#1417](https://github.com/microsoft/onnxruntime/issues/1417) back in July, 2019, I did not have to address dynamic shape inputs so something has changed.  I am unclear if dynamic shape inputs are necessary for this use-case and if so, is it something I handle when building the model in ML.Net, converting the ML.Net model to onnx, or when assigning the dense tensor to optimize inference performance.

**Comment by harishsk** [comment]
Hi @whitmark,
Back in July 2019 there was very minimal onnx export functionality in ML.NET. While not still complete, it is a lot better today. We have added export support for many more transformers with support for more data types. 
While exporting a model the batch size needed for inferencing is not always known. To help with this, Onnx supports specifying inputs with dynamic shapes. Back in July last year, a model exported from ML.NET wouldn't have this support.  At the request of other customers, we have added this feature to export models that accept inputs of different sizes. ML.NET now always exports models with dynamic shapes for input. But you do not have to worry about it as long as you specify the correct valid dimensions for your input.

**Comment by whitmark** [comment]
Thanks @harishsk for the background explanation.  Sorry if I am being a bit dense in my understanding of tensor operations (pun intended).  I am looking to inference an ML.net model in a .Net Framework 4.8 app, hence the need for the conversion to onnx.  I would imagine the use of tabular 1D ML.net models is a common use-case and in the absence of a relevant example in the docs or test scripts, can you offer sample code that would 1) be selective of input columns so I can skip reference columns and 2) assign the proper tensor shape/dims if I intend to inference one record at a time (batch=1?) given the target framework.

**Comment by whitmark** [comment]
The binary classification model, trained with ML.Net using a Lightgbm trainer, and saved in onnx format will present a dynamic shape dimensions of [-1, 1] for each column so I found I can load a container list with names and values for inferencing using dimensions of [1,1] as shown in the code snippet below:

```
     var tensor = new DenseTensor<float>(new float[] { inputData1[i - startSlot] }, new int[] { 1, 1 });
     container.Add(NamedOnnxValue.CreateFromTensor<float>(name, tensor));
```

Furthermore, after some trial and error, I clarified a few things:

1. To run an inference, the inference container columns (100) and data types need to match the model's InferenceSession inputMetadata columns (100) and datatype types.   In hindsight, this seems obvious, but last year I suspect only float columns were supported with ML.net onnx models so other data type columns  were ignored so I could get away with just submitting float values.

2. Of the 100 input columns, 3 columns are  for reference (2 string, 1 int64), 1  column is for the label (1 bool) and remaining 96 columns are feature values (96 float).  When the model was saved as an onnx file, at first I was not expecting the 3 reference columns to be retained since they were dropped in the training pipeline nor did I expect the label column to be retained either but I can deal with both being included.

3. Although the ML.Net model uses a float value  of 0 or 1 for the binary classification label, the onnx model saves it as a bool type so the inference type for the label column must also be bool.

I am good to go for now but it would be nice to to have a working sample for others, that include saving tabular data-driven ML.Net model and retaining reference columns that are not labels or features.  Thanks for the clues along the way.
